pkg/validation/validation_test.go

```diff
@@ -398,3 +404,397 @@ func TestGatewayClass(t *testing.T) {
+						{
+							Port:     gatewayv1alpha1.PortNumber(int32(2)),
+							Protocol: gatewayv1alpha1.HTTPProtocolType,
+						},
+					},
+				},
+			},
+			expected: false,
+		},
+		"invalid wildcard host gateway": {
+			gw: &gatewayv1alpha1.Gateway{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: ns.Name,
+					Name:      "invalid-wildcard-host-gateway",
+				},
+				Spec: gatewayv1alpha1.GatewaySpec{
+					GatewayClassName: gc.Name,
+					Listeners: []gatewayv1alpha1.Listener{
+						{
+							Hostname: &invalidWildcard,
+							Port:     gatewayv1alpha1.PortNumber(int32(1)),
+							Protocol: gatewayv1alpha1.HTTPSProtocolType,
+						},
+						{
+							Port:     gatewayv1alpha1.PortNumber(int32(2)),
+							Protocol: gatewayv1alpha1.HTTPProtocolType,
+						},
+					},
+				},
+			},
+			expected: false,
+		},
+		"invalid gatewayclass reference": {
+			gw: &gatewayv1alpha1.Gateway{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: ns.Name,
+					Name:      "invalid-gatewayclass-reference",
+				},
+				Spec: gatewayv1alpha1.GatewaySpec{
+					GatewayClassName: "nonexistent",
+					Listeners: []gatewayv1alpha1.Listener{
+						{
+							Port:     gatewayv1alpha1.PortNumber(int32(1)),
+							Protocol: gatewayv1alpha1.HTTPSProtocolType,
+						},
+						{
+							Port:     gatewayv1alpha1.PortNumber(int32(2)),
+							Protocol: gatewayv1alpha1.HTTPProtocolType,
+						},
+					},
+				},
+			},
+			expected: false,
+		},
+		"invalid number of gateway listeners": {
+			gw: &gatewayv1alpha1.Gateway{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: ns.Name,
+					Name:      "invalid-number-of-gateway-listeners",
+				},
+				Spec: gatewayv1alpha1.GatewaySpec{
+					GatewayClassName: gc.Name,
+					Listeners: []gatewayv1alpha1.Listener{
+						{
+							Port:     gatewayv1alpha1.PortNumber(int32(1)),
+							Protocol: gatewayv1alpha1.HTTPSProtocolType,
+						},
+					},
+				},
+			},
+			expected: false,
+		},
+		"duplicate listener port numbers": {
+			gw: &gatewayv1alpha1.Gateway{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: ns.Name,
+					Name:      "duplicate-listener-port-numbers",
+				},
+				Spec: gatewayv1alpha1.GatewaySpec{
+					GatewayClassName: gc.Name,
+					Listeners: []gatewayv1alpha1.Listener{
+						{
+							Port:     gatewayv1alpha1.PortNumber(int32(1)),
+							Protocol: gatewayv1alpha1.HTTPSProtocolType,
+						},
+						{
+							Port:     gatewayv1alpha1.PortNumber(int32(1)),
+							Protocol: gatewayv1alpha1.HTTPProtocolType,
+						},
+					},
+				},
+			},
+			expected: false,
+		},
+		"invalid listener protocol": {
+			gw: &gatewayv1alpha1.Gateway{
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: ns.Name,
+					Name:      "invalid-listener-protocol",
+				},
+				Spec: gatewayv1alpha1.GatewaySpec{
+					GatewayClassName: gc.Name,
+					Listeners: []gatewayv1alpha1.Listener{
+						{
+							Port:     gatewayv1alpha1.PortNumber(int32(1)),
+							Protocol: gatewayv1alpha1.TLSProtocolType,
```

Comment:
i think this is allowed by Contour itself now, see https://github.com/projectcontour/contour/commit/72edf8beb1ebbc5dc17319b92b15fc474c85de30